### PR TITLE
aws: fork mbedtls 2.25.0 base64 utility to avoid 2.26.0+ performance hit

### PR DESCRIPTION
**Pull request by matthewfala**
Migrate to performant 2.25.0-mbedtls base64 implementation.

Tested via unit tests + 1 end to end test
Sent logs to firehose via the out_kinesis_firehose plugin which utilizes the implementation.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [N/A] Debug log output from testing the change

- [N/A] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found (No memory allocations)

**Documentation**

- [ ] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Review by PettitWesley — APPROVED** [review]
@edsiper I'd like your review/approval on this as well.

**Comment by edsiper** [comment]
@PettitWesley base64 unit test is failing

https://github.com/fluent/fluent-bit/runs/4478384757?check_suite_focus=true#step:5:1810

**Comment by edsiper** [comment]
Why not to make it generic use instead of just _aws_ scoped? I think it could be useful as flb_base64.c alone

**Comment by matthewfala** [comment]
Strange how it passes on some operating systems. Grateful for the crossplatform unit test runner. I'll take a look. I can refactor this to flb_base64

Since I did some refactoring to the aws unit test structure for the base64 test, I might switch to an aws placeholder test that noops, so future tests to aws can be added easily if that is alright.

**Comment by matthewfala** [comment]
@PettitWesley Could you take a look at the changes? https://github.com/fluent/fluent-bit/pull/4422/commits/3c08a687820ee1c505347f3352be0933816a6829

I think I resolved the problem with the testcase, it may have just been a lack of a null character when calling strlen. I initialize the array to 0 now in the decode test case.

**Comment by PettitWesley** [comment]
@matthewfala CI says the tests are still failing? I am good to approve/merge once everything passes.

**Comment by matthewfala** [comment]
Need to refactor the commit messages, then this should be done

**Review by PettitWesley — APPROVED** [review]

**Comment by matthewfala** [comment]
@edsiper Could you please take a look at this when you get the chance?

**Comment by nokute78** [comment]
@matthewfala  @PettitWesley  @krispraws 
Before revewing
- Could you fix conflict ?
- How much did that improve ?
- How about upgrading mbedtls v2.28.0 since the patch https://github.com/ARMmbed/mbedtls/pull/4835 has been merged ? https://github.com/ARMmbed/mbedtls/commit/8e8251a6fa065e34aa033bd0ecb526506923c8eb

**Comment by edsiper** [comment]
thanks @nokute78 for the hints about mbedtls, I will make sure to upgrade in the 1.8 branch

**Comment by matthewfala** [comment]
Hi @nokute78,
Actually, could you let me know if this pr will be considered? Then I'll fix the conflicts.

According to Ramya's testing as described in the following issue:
https://github.com/fluent/fluent-bit/issues/4110
The performance hits of Mbedtls's new base64 function is severe, causing significant drop in throughput in our plugin that uses mbedtls.

It appears the newest base64 implementation is still slower than the older implementation as described by the changelog message in `ChangeLog.d/base64-ranges.txt`
```
Changes
   * Improve the performance of base64 constant-flow code. The result is still
     slower than the original non-constant-flow implementation, but much faster
     than the previous constant-flow implementation. Fixes #4814.
```

Would forking be considered above just upgrading mbedtls?

**Comment by nokute78** [comment]
@matthewfala In my opinion, forking can be a risk and take a porting cost. e.g. Watching upstream and merging patches.
If the performance using mbedtls v2.28 is not acceptable for user/developer, we should merge this PR as an interim patch.
If it is not, we need to upgrade mbedtls.

How much slower is it using mbedtls v2.28 ?

**Comment by matthewfala** [comment]
Thank you @nokute78. I like the interim patch idea. I'll fix the merge conflicts. We can assess the speed difference after, though I am not familiar with profiling, so I'll either learn or ask someone else on the team do help.

**Comment by matthewfala** [comment]
@nokute78 Could you take a look? Resolved merge conflicts

**Review by nokute78 — CHANGES_REQUESTED** [review]
Could you add test case to compare performance between `flb_base64_*` and `mbedtls_base64_*` ?
The test fails when `mbedtls_base64_*` are faster than `flb_base64_*` or the performance of `mbedtls_base64_*` becomes acceptable.

We can detect  it is the time to drop the interim patch and back to mbedtls.

File: tests/internal/base64.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            elapsed_time_flb / 1000000L, elapsed_time_flb % 1000000L);
+    printf("flb_mbedtls_encode %d iterations complete in %lums and %luns\n", iter,
+            elapsed_time_mbedtls / 1000000L, elapsed_time_mbedtls % 1000000L);
+
+    TEST_CHECK(elapsed_time_flb * acceptableTimeIncrease < elapsed_time_mbedtls);
+    TEST_MSG("Base64 mbedtls performance is now similar to fluent bit's. "
+             "Revert to mbedtls base64.");
+    return;
+}
+
```
Comment (matthewfala) [inline tests/internal/base64.c:89]: Added performance test. On my machine, I get the following result
```
flb_base64_encode 100000 iterations complete in 3ms and 923342ns
flb_mbedtls_encode 100000 iterations complete in 943ms and 308068ns
[ OK ]
```

Comment (matthewfala) [reply tests/internal/base64.c:89]: The code is triggered to fail when mbedtls is faster than flb_base64 slowed down by 20%

Comment (nokute78) [reply tests/internal/base64.c:89]: Thank you!

**Review by nokute78 — APPROVED** [review]
@matthewfala Thank you for updating.
LGTM.

**Comment by matthewfala** [comment]
@nokute78 YW. Could you give me an eta on when this will be merged?

**Comment by nokute78** [comment]
@matthewfala I can't merge since "Merging can be performed automatically with 1 approving review." .
@edsiper Could you merge this ?

**Comment by edsiper** [comment]
@matthewfala  would you please submit a PR for 1.8 branch ?

**Comment by matthewfala** [comment]
Definitely! Will publish a pr to 1.8 today.

**Pull request merged**